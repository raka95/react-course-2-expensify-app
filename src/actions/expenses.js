import uuid from "uuid";
import database from "../firebase/firebase";
//Add-Expence



export const addExpense=(expense)=>({
    type:"Add-EXPENSE",
    expense
}
);

export const startAddExpense=(expenseData={})=>{
    return (dispatch)=>{
        const { 
            description='',
            note='',
            amount=0,
            createdAt=0
        }=expenseData;
        const expense={description,note,amount,createdAt};

        // return for test v153
        return database.ref("Expenses").push(expense).then((ref)=>{
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        })
    }
}

// export const addExpense=(
//     {
//         description='',
//         note='',
//         amount=0,
//         createdAt=0
//     }={}
// )=>({
//     type:"Add-EXPENSE",
//     expense:{
//         id:uuid(),
//         description,
//         note,
//         amount,
//         createdAt
//     }
// }
// );



//Remove-Expense

export const removeExpense=({id})=>(
    {
        type:"Remove-Expense",
        id
    }
);

export const startRemoveExpense=({id}={})=>{
    return (dispatch)=>{
            return database.ref(`Expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense({id}));
        });
    }
}
//Edit-Expense

export const editExpense=(id,update)=>({
    type:"Edit-Expense",
    id,
    update
});

//SET-EXPENSE

export const setExpense=(expenses)=>({
    type:"SET-EXPENSES",
    expenses
})

//export startSetExpense
export const startSetExpense=()=>{
    return(dispatch)=>{
        return database.ref('Expenses').once('value').then((snapshot)=>{
            const expenses=[];            
            snapshot.forEach((childSnapshot)=>{
                expenses.push({
                    ...childSnapshot.val(),
                    id:childSnapshot.key,
                });
            });
           dispatch(setExpense(expenses));
        });
    }
}