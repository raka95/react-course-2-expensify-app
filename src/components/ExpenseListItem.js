import React from 'react';
import { NavLink } from 'react-router-dom';


export const ExpenseListItem = ({ description, amount, createdAt, id, dispatch }) => (
    <div>
        <NavLink to={`/edit/${id}`} >
            <h1>{description}</h1>
        </NavLink>
        <p>amount: {amount} - createdAt: {createdAt}</p>
    </div>
);

export default (ExpenseListItem);
// const ExpenseListItem=({description,amount,createdAt})=>(
//     <div>
//         <h1>{description}</h1>
//         <p>amount: {amount} - createdAt: {createdAt}</p>
//     </div>
// );
// export default ExpenseListItem;