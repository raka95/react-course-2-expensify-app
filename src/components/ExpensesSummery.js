import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from "../selectors/expenses";
import selectExpensesTotal from '../selectors/expenses-total';


export const ExpensesSummery=({expense})=>(
    <div>
    <h1>Viewing {expense.length} expenses totalling {selectExpensesTotal(expense)}</h1>
    </div>
);
//{
// //  return expense.reduce((a,b)=>{
// //      return a + b.amount;
// //  },0);
// return expense.map((val)=>val.amount).reduce((sum,value)=>{sum+value,0});
// };

const mapStateToProps=(state)=>({
    expense:selectExpenses(state.expenses,state.filters)
})
export default connect(mapStateToProps)(ExpensesSummery);
