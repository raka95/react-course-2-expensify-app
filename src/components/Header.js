import {Link,NavLink} from  'react-router-dom';
import React from 'react';
const Header=()=>(
    <header>
    <h1>Expensify</h1>
    <NavLink to='/' activeClassName="is-active" exact={true}>Home </NavLink>
    <NavLink to='/create' activeClassName="is-active">Create </NavLink> 
    </header>
);
export default Header;
//   <NavLink to='/help' activeClassName="is-active">Help </NavLink> 