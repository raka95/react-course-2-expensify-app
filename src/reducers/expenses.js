//Expences Reducers  

const expensesReducerDefault=[];
export default (state=expensesReducerDefault,action)=>{
// const expensesReducer=(state=expensesReducerDefault,action)=>{
    switch(action.type){
        case "Add-EXPENSE":
            return [
                ...state,
                action.expense
            ];
        case "Remove-Expense":
            return state.filter(({id})=>id!==action.id);
        case "Edit-Expense":
            return state.map((val)=>{
                if(val.id===action.id)
                    return {
                        ...val,
                        ...action.update
                    };
                    return val;
            })
        default:
            return state;
    }
};
//export default expensesReducer;