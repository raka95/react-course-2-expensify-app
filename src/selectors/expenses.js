// get visible expenses
export default (expenses,{text,sortBy,startDate,endDate})=>{
// const getVisibleExpenses=(expenses,{text,sortBy,startDate,endDate})=>{
    return expenses.filter((val)=>{
     const createdAtMoment=val.createdAt;
     const startDateMAtch=startDate ? startDate.isSameOrBefore(createdAtMoment,"day") :true ;
     const endDateMatch=endDate ?  endDate.isSameOrAfter(createdAtMoment,"day"):true ;
     const textMatch=val.description.toLowerCase().includes(text);
 
     return startDateMAtch&&endDateMatch&&textMatch;
    }).sort((a,b)=>{
        if(sortBy==="date")
         return b.createdAt-a.createdAt;
         else if(sortBy==="amount")
         return b.amount-a.amount;
    })
 };
 