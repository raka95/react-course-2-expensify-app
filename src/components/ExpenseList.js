import React from 'react';
import {connect} from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";


export const ExpenseList=({expense})=>(
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>        
        </div>
        <div className="list-body">
            {expense.length===0 ? (
                <div className="list-item list-item--message">
                <span>No Expenses</span>
                </div>
            ):(
                expense.map((val)=>(
                    <ExpenseListItem 
                        key={val.id} 
                        {...val}/>
                    )
                )
            )}
        </div>    
    </div>
);
//
// 
const mapStateToProps=(state)=> (
    {
        expense:selectExpenses(state.expenses,state.filters)
    }
);
export default connect(mapStateToProps)(ExpenseList);
//

// const ConnectExpenseList=connect((state)=>{
//     return {
//         name:"Ramin",
//         expense:state.expenses,
//         filters:state.filters
//     };
// })(ExpenseList);
// export default ConnectExpenseList;

