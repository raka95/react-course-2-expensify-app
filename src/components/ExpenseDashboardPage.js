// import {Link,NavLink} from  'react-router-dom';
import React from 'react';
import ConnectExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummery from './ExpensesSummery';

export const ExpenseDashboardPage=()=>(
    <div>
        <ExpensesSummery />
        <ExpenseListFilters />
        <ConnectExpenseList />
    </div>
);
export default ExpenseDashboardPage;