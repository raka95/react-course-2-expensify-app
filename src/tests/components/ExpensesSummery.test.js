import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummery} from '../../components/ExpensesSummery';
import expenses from '../fixtures/expenses';

test("should render expensesLength in expenseSummery",()=>{
    const wrapper = shallow(<ExpensesSummery expense={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

// test("should render expensesLength in expenseSummery",()=>{
//     const wrapper = shallow(<ExpensesSummery expense={expenses}/>);
//     expect(wrapper).toMatchSnapshot();
// });