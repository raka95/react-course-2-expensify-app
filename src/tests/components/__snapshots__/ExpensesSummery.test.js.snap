// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expensesLength in expenseSummery 1`] = `
<div>
  <h1>
    Viewing 
    3
     expenses totalling 
    2590
  </h1>
</div>
`;
